File: ios/KairoWidgetExtension/Data/KairoWidgetData.swift
// KairoWidgetData.swift
// Modelo + leitura do App Group + lógica pura de truncamento e streak.
// A verdade é escrita pelo app principal (ver KairoWidgetBridge no Runner).

import Foundation

struct KairoWidgetData {
    var phrase: String
    var nextPracticeTitle: String
    var nextPracticeTime: Date?
    var streakCount: Int
    var lastUpdated: Date?

    /// Sem nenhum dado útil ainda (fresh install) → estado "回路 + Kairo".
    var isEmpty: Bool { phrase.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }

    /// Dados defasados (> 24h) → mostrar ponto discreto de "não fresco".
    var isStale: Bool {
        guard let last = lastUpdated else { return false }
        return Date().timeIntervalSince(last) > 24 * 60 * 60
    }

    static let vazio = KairoWidgetData(
        phrase: "", nextPracticeTitle: "", nextPracticeTime: nil,
        streakCount: 0, lastUpdated: nil
    )
}

enum KairoStore {
    static let appGroup = "group.com.kairo.shared"

    static func ler() -> KairoWidgetData {
        guard let d = UserDefaults(suiteName: appGroup) else { return .vazio }
        let phrase = d.string(forKey: "phrase_today") ?? ""
        let title  = d.string(forKey: "next_practice_title") ?? ""
        let timeISO = d.string(forKey: "next_practice_time")
        let streak = d.integer(forKey: "streak_count")
        let updatedEpoch = d.double(forKey: "last_updated")

        var time: Date? = nil
        if let iso = timeISO, !iso.isEmpty {
            time = ISO8601DateFormatter().date(from: iso)
        }
        let updated: Date? = updatedEpoch > 0 ? Date(timeIntervalSince1970: updatedEpoch) : nil

        return KairoWidgetData(
            phrase: phrase, nextPracticeTitle: title, nextPracticeTime: time,
            streakCount: streak, lastUpdated: updated
        )
    }
}

// ── Lógica pura (espelhada em Dart/Kotlin e coberta por testes) ──────────────

enum KairoFormat {
    /// Trunca a frase com EM-DASH (nunca com "..."). Corta na borda de palavra
    /// quando possível para não cortar no meio de uma palavra.
    /// (Reconcilia a nota do inline com a regra global "never '...'": usamos —.)
    static func truncar(_ texto: String, max: Int) -> String {
        let limpo = texto.trimmingCharacters(in: .whitespacesAndNewlines)
        guard limpo.count > max, max > 1 else { return limpo }
        let corte = limpo.prefix(max - 1)
        // recua até o último espaço para não partir uma palavra
        if let ultimoEspaco = corte.range(of: " ", options: .backwards) {
            let palavra = corte[..<ultimoEspaco.lowerBound]
            let semPontuacao = palavra.trimmingCharacters(in: CharacterSet(charactersIn: " ,;:"))
            if semPontuacao.count >= max / 2 {
                return semPontuacao + "—"
            }
        }
        return corte.trimmingCharacters(in: .whitespaces) + "—"
    }

    /// Quantos stones desenhar e se há overflow (> 7 → 6 stones + bullet).
    static func stones(streak: Int) -> (cheios: Int, overflow: Bool) {
        if streak <= 0 { return (0, false) }
        if streak <= KairoBrand.maxStones { return (streak, false) }
        return (KairoBrand.maxStones - 1, true) // 6 + "·"
    }
}

File: android/app/src/main/kotlin/com/example/kairo/widget/data/KairoWidgetData.kt
package com.example.kairo.widget.data

import android.content.Context
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.intPreferencesKey
import androidx.datastore.preferences.core.longPreferencesKey
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.runBlocking

// DataStore (Preferences) "kairo_widget" — fonte de dados dos widgets, escrita
// pelo app principal (ver MainActivity MethodChannel) e lida pelos providers.
val Context.kairoWidgetStore by preferencesDataStore(name = "kairo_widget")

object KairoKeys {
    val phrase = stringPreferencesKey("phrase_today")
    val practiceTitle = stringPreferencesKey("next_practice_title")
    val practiceTime = stringPreferencesKey("next_practice_time") // ISO8601, opcional
    val streak = intPreferencesKey("streak_count")
    val lastUpdated = longPreferencesKey("last_updated") // epoch ms
}

data class KairoWidgetData(
    val phrase: String = "",
    val practiceTitle: String = "",
    val practiceTime: String? = null,
    val streak: Int = 0,
    val lastUpdated: Long = 0L,
) {
    val isEmpty: Boolean get() = phrase.trim().isEmpty()
    val isStale: Boolean
        get() = lastUpdated > 0 && System.currentTimeMillis() - lastUpdated > 24L * 60 * 60 * 1000
}

object KairoStore {
    /// Leitura síncrona (via runBlocking) — chamada de dentro do onUpdate do
    /// AppWidgetProvider, que já roda fora da main thread no broadcast.
    fun ler(context: Context): KairoWidgetData = runBlocking {
        val p: Preferences = context.kairoWidgetStore.data.first()
        KairoWidgetData(
            phrase = p[KairoKeys.phrase] ?: "",
            practiceTitle = p[KairoKeys.practiceTitle] ?: "",
            practiceTime = p[KairoKeys.practiceTime]?.takeIf { it.isNotEmpty() },
            streak = p[KairoKeys.streak] ?: 0,
            lastUpdated = p[KairoKeys.lastUpdated] ?: 0L,
        )
    }

    suspend fun gravar(
        context: Context,
        phrase: String?,
        practiceTitle: String?,
        practiceTime: String?,
        streak: Int?,
    ) {
        context.kairoWidgetStore.edit { p ->
            phrase?.let { p[KairoKeys.phrase] = it }
            practiceTitle?.let { p[KairoKeys.practiceTitle] = it }
            // practiceTime: string vazia/null remove a chave.
            if (practiceTime.isNullOrEmpty()) p.remove(KairoKeys.practiceTime)
            else p[KairoKeys.practiceTime] = practiceTime
            streak?.let { p[KairoKeys.streak] = it }
            p[KairoKeys.lastUpdated] = System.currentTimeMillis()
        }
    }
}

// ── Lógica pura (espelhada em Swift/Dart e coberta por testes) ───────────────
object KairoFormat {
    const val MAX_STONES = 7

    /// Trunca com EM-DASH (nunca "..."). Recua à borda de palavra quando possível.
    fun truncar(texto: String, max: Int): String {
        val limpo = texto.trim()
        if (limpo.length <= max || max <= 1) return limpo
        val corte = limpo.substring(0, max - 1)
        val ultimoEspaco = corte.lastIndexOf(' ')
        if (ultimoEspaco >= max / 2) {
            return corte.substring(0, ultimoEspaco).trimEnd(' ', ',', ';', ':') + "—"
        }
        return corte.trimEnd() + "—"
    }

    /// (stones cheios, overflow?) — > 7 vira 6 stones + "·".
    fun stones(streak: Int): Pair<Int, Boolean> = when {
        streak <= 0 -> 0 to false
        streak <= MAX_STONES -> streak to false
        else -> (MAX_STONES - 1) to true
    }
}
